chain/sub/incoming.go

```diff
@@ -121,34 +144,206 @@ func (bv *BlockValidator) flagPeer(p peer.ID) {
 	}
 
+	// we want to ensure that it is a block from a known miner; we reject blocks from unknown miners
+	// to prevent spam attacks.
+	// the logic works as follows: we lookup the miner in the chain for its key.
+	// if we can find it then it's a known miner and we can validate the signature.
+	// if we can't find it, we check whether we are (near) synced in the chain.
+	// if we are not synced we cannot validate the block and we must ignore it.
+	// if we are synced and the miner is unknown, then the block is rejcected.
+	key, err := bv.getMinerWorkerKey(ctx, blk)
+	if err != nil {
+		if bv.isChainNearSynced() {
+			log.Warnf("received block message from unknown miner over pubsub; rejecting message")
+			recordFailure("unknown_miner")
+			return pubsub.ValidationReject
+		} else {
+			log.Warnf("cannot validate block message; unknown miner in unsynced chain")
+			return pubsub.ValidationIgnore
+		}
+	}
+
+	blob, err := blk.Header.SigningBytes()
+	if err != nil {
+		// this shouldn't happen; it's an encoding error of the object we just decoded sans the
+		// signature...
+		log.Errorf("error retrieving signing bytes for block header: %s", err)
+		return pubsub.ValidationIgnore
+	}
+
+	err = sigs.Verify(blk.Header.BlockSig, key, blob)
+	if err != nil {
+		log.Errorf("block signature verification failed: %s", err)
+		recordFailure("signature_verification_failed")
+		return pubsub.ValidationReject
+	}
+
+	// it's a good block! make sure we've only seen it once
 	if bv.recvBlocks.add(blk.Header.Cid()) > 0 {
 		// TODO: once these changes propagate to the network, we can consider
 		// dropping peers who send us the same block multiple times
 		return pubsub.ValidationIgnore
 	}
 
+	// all good, accept the block
 	msg.ValidatorData = blk
 	stats.Record(ctx, metrics.BlockValidationSuccess.M(1))
 	return pubsub.ValidationAccept
 }
 
+func (bv *BlockValidator) isOldBlock(msg *types.BlockMsg) bool {
+	cutoff := time.Now().UnixNano() - int64(2*time.Minute)
+	if msg.Header.Timestamp < uint64(cutoff) {
+		return true
+	}
+	return false
+}
+
+func (bv *BlockValidator) isChainNearSynced() bool {
+	ts := bv.chain.GetHeaviestTipSet()
+	timestamp := ts.MinTimestamp()
+	now := time.Now().UnixNano()
+	cutoff := uint64(now) - uint64(6*time.Hour)
+	return timestamp > cutoff
+}
+
+func (bv *BlockValidator) validateMsgMeta(ctx context.Context, msg *types.BlockMsg) error {
+	var bcids, scids []cbg.CBORMarshaler
+	for _, m := range msg.BlsMessages {
+		c := cbg.CborCid(m)
+		bcids = append(bcids, &c)
+	}
+
+	for _, m := range msg.SecpkMessages {
+		c := cbg.CborCid(m)
+		scids = append(scids, &c)
+	}
+
+	// TODO there has to be a simpler way to do this without the blockstore dance
+	bs := cbor.NewCborStore(bstore.NewBlockstore(dstore.NewMapDatastore()))
+
+	bmroot, err := amt.FromArray(ctx, bs, bcids)
+	if err != nil {
+		return err
+	}
+
+	smroot, err := amt.FromArray(ctx, bs, scids)
+	if err != nil {
+		return err
+	}
+
+	mrcid, err := bs.Put(ctx, &types.MsgMeta{
```

Comment:
You can just create a cbor ipld node, and get its cid

Reply:
any pointers for this? I would like to get rid of the use of a temporary datastore.

Reply:
Not sure if we have a better util for this currently - https://github.com/ipfs/go-ipld-cbor/blob/master/store.go#L70-L99

Reply:
ok, duplicating that code is not something I want to do; when we get an actual utility function I will gladly repace the temporary datastore code  (there is already a TODO :)